cpp/include/cuvs/neighbors/ann_mg.hpp

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <nccl.h>
+#include <raft/core/device_resources.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <cuvs/neighbors/cagra.hpp>
+#include <cuvs/neighbors/ivf_flat.hpp>
+#include <cuvs/neighbors/ivf_pq.hpp>
+
+/**
+ * @brief Error checking macro for NCCL runtime API functions.
+ *
+ * Invokes a NCCL runtime API function call, if the call does not return ncclSuccess, throws an
+ * exception detailing the NCCL error that occurred
+ */
+#define RAFT_NCCL_TRY(call)                        \
+  do {                                             \
+    ncclResult_t const status = (call);            \
+    if (ncclSuccess != status) {                   \
+      std::string msg{};                           \
+      SET_ERROR_MSG(msg,                           \
+                    "NCCL error encountered at: ", \
+                    "call='%s', Reason=%d:%s",     \
+                    #call,                         \
+                    status,                        \
+                    ncclGetErrorString(status));   \
+      throw raft::logic_error(msg);                \
+    }                                              \
+  } while (0);
+
+namespace cuvs::neighbors::mg {
+enum parallel_mode { REPLICATED, SHARDED };
+}
+
+namespace cuvs::neighbors::ivf_flat {
+struct mg_index_params : cuvs::neighbors::ivf_flat::index_params {
+  cuvs::neighbors::mg::parallel_mode mode;
+};
+}  // namespace cuvs::neighbors::ivf_flat
+
+namespace cuvs::neighbors::ivf_pq {
+struct mg_index_params : cuvs::neighbors::ivf_pq::index_params {
+  cuvs::neighbors::mg::parallel_mode mode;
+};
+}  // namespace cuvs::neighbors::ivf_pq
+
+namespace cuvs::neighbors::cagra {
+struct mg_index_params : cuvs::neighbors::cagra::index_params {
+  cuvs::neighbors::mg::parallel_mode mode;
+};
+}  // namespace cuvs::neighbors::cagra
+
+namespace cuvs::neighbors::mg {
+using pool_mr = rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>;
+
+struct nccl_clique {
+  nccl_clique(const std::vector<int>& device_ids);
+  const raft::device_resources& set_current_device_to_root_rank() const;
+  ~nccl_clique();
+
+  int root_rank_;
+  int num_ranks_;
+  std::vector<int> device_ids_;
+  std::vector<ncclComm_t> nccl_comms_;
```

Comment:
We will have to forward declare `ncclComm_t` so that we can avoid including NCCL header:
```
class ncclComm_t {};

```suggestion
  std::vector<std::unique_ptr<ncclComm_t>> nccl_comms_;
```